src/cards/polymorph.ts

```diff
@@ -0,0 +1,194 @@
+      toSourceUnit.id,
+      fromUnit.x,
+      fromUnit.y,
+      fromUnit.faction,
+      toSourceUnit.info.image,
+      UnitType.AI,
+      toSourceUnit.info.subtype,
+      { ...toSourceUnit.unitProps, isMiniboss: fromUnit.isMiniboss, originalLife: fromUnit.originalLife },
+      underworld,
+      prediction
+    );
+
+    if (unit != undefined) {
+      // Keep Modifiers
+      for (const modifierKey of Object.keys(unit.modifiers)) {
+        const modifier = allModifiers[modifierKey];
+        const modifierInstance = unit.modifiers[modifierKey];
+        if (modifier && modifierInstance) {
+          if (modifier?.add) {
+            modifier.add(unit, underworld, prediction, modifierInstance.quantity, modifierInstance);
+          }
+        } else {
+          console.error("Modifier doesn't exist? This shouldn't happen.");
+        }
+      }
+
+      // Cleanup old unit and remove it from targets
+      Unit.cleanup(fromUnit, false);
+    }
+    return unit;
+  } else {
+    // Only change vfx/sfx for player
+    fromUnit.defaultImagePath = toSourceUnit.unitProps.defaultImagePath || fromUnit.defaultImagePath;
+    fromUnit.animations = toSourceUnit.animations || fromUnit.animations;
+    fromUnit.sfx = toSourceUnit.sfx || fromUnit.sfx;
+    fromUnit.bloodColor = toSourceUnit.unitProps.bloodColor || fromUnit.bloodColor;
+
+    Image.changeSprite(
+      fromUnit.image,
+      fromUnit.animations.idle,
+      fromUnit.image?.sprite.parent,
+      undefined,
+    );
+
+    return fromUnit;
+  }
+}
+
+function polymorphPickup(fromPickup: IPickup, underworld: Underworld, prediction: boolean, toPickupSource?: Pickup.IPickupSource): IPickup | undefined {
+  // If a specific ID isn't passed in, choose a random one
+  if (!toPickupSource) {
+    // Don't polymorph purple portals
+    if (fromPickup.name == Pickup.PORTAL_PURPLE_NAME) return undefined;
+
+    let possiblePickupTypes = pickups.filter(p => isModActive(p, underworld) && p.name != fromPickup.name
+      && p.name != Pickup.PORTAL_PURPLE_NAME && p.name != Pickup.RECALL_POINT);
+
+    // We have to seed this to prevent multiplayer desync
+    // ID is required here to ensure that every polymorph changes the seed to prevent looping
+    const seed = seedrandom(`${getUniqueSeedString(underworld)} - ${fromPickup.id}`);
+    toPickupSource = chooseOneOfSeeded(possiblePickupTypes, seed);
+    if (toPickupSource == undefined) {
+      console.error("Polymorph failed to choose a new pickup type.");
+      return undefined;
+    }
+  }
+
+  console.log(toPickupSource);
+  const pickup = Pickup.create({ pos: fromPickup, pickupSource: toPickupSource, logSource: 'spawnPickup' }, underworld, prediction);
+  if (pickup != undefined) {
+    if (!prediction) {
+      playSFXKey('spawnPotion');
+      floatingText({ coords: fromPickup, text: toPickupSource.name });
+    }
+    // Cleanup old pickup and remove it from targets
+    Pickup.removePickup(fromPickup, underworld, prediction);
```

Comment:
This line doesn't remove the pickup from targets, that is handled in the calling function right now. Let's pass the effectState into this function and handle the cleanup and adding the new pickup to the target array here